test-data/unit/check-custom-pyproject-plugin.test

```diff
@@ -0,0 +1,737 @@
+-- Test cases for user-defined plugins
```

Comment:
This and the following test files are pretty verbose. I think that it's better to try to keep the test files shorter so that each test covers some additional toml/config feature. A large number of tests slows local test runs and adds maintenance overhead.

It's good that you added so many tests, as it adds confidence to the implementation. Once this PR is close to ready, I'd recommend removing tests that are unlikely to increase test coverage significantly.

Reply:
> It's good that you added so many tests, as it adds confidence to the implementation. Once this PR is close to ready, I'd recommend removing tests that are unlikely to increase test coverage significantly.

I agree. I did these mostly to ensure that I covered as many use cases as I could, but realize they are a bit out of control. I'm definitely open to paring them down and/or combining a lot of the positive unit tests.